### PR TITLE
Handle query_channel_range gossip queries

### PR DESCRIPTION
**Pull request by bmancini55**
Finally back in the saddle on this. This PR implements the `NetGraphMsgHandler` method for processing inbound `query_channel_range` messages. I'm working on another PR to handle `query_short_channel_ids` as well.

Notes:

- This implementation enqueues each reply message as an event. Right now, at most this will only generate a few events per query due to the graph size. In the future it could generate a large number of events if the graph is large. I'm considering switching to a polling model like the `init_sync` code in  `do_attempt_write_data` to be more aware of the write buffer. Wanted opinions before I go down that path though.
- Sequencing metadata in replies (`first_blocknum`, `number_of_blocks`) is still necessary to be spec compliant even though its ugly. I added notes on overflow safety. I'm happy to make that more explicit if that's preferable.

### CODE REVIEW

File: lightning/src/routing/network_graph.rs
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:378]: nit: cant we just cheat and always send `msg.first_blocknum`? There's nothing that says we cant send 0, and a receiver needs to parse individual values to figure out the real block anyway.

File: lightning/src/routing/network_graph.rs
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:372]: Instead of `batch_index` you can replace the for loop with `for (batch, batch_index) in batches.into_iter().enumerate() {` (or maybe the args are swapped).

**Review by TheBlueMatt — APPROVED** [review]
Some pretty trivial nits, but overall looks good. Will get a second reviewer this week.

> This implementation enqueues each reply message as an event. Right now, at most this will only generate a few events per query due to the graph size. In the future it could generate a large number of events if the graph is large. I'm considering switching to a polling model like the init_sync code in do_attempt_write_data to be more aware of the write buffer. Wanted opinions before I go down that path though.

Hmm, right, ~5 messages for ~32KB is definitely not worth doing that for, but we will need it to respond to `ReplyChannelRange` messages, which could be a lot of additional messages and we can't block channel messages while sending them. For that we definitely need to pull the InitSyncTracker struct out into `network_graph.rs` and make it an associated type for the `RoutingMessageHandler` trait. I leave it up to you if you want to take that complexity here and also apply it to this code or not.

File: lightning/src/ln/msgs.rs
Comment (bmancini55) [reply lightning/src/ln/msgs.rs:1693]: Added constants to scid_utils, however in this spot, we need uint32 max instead of the `0xff_ff_ff` constant in the scid calcs to enable reply sequencing.

File: lightning/src/util/scid_utils.rs
Comment (bmancini55) [reply lightning/src/util/scid_utils.rs]: Done in dd0bb256940d885a1e8dfd3ae8be6e5ea648944e

File: lightning/src/routing/network_graph.rs
Comment (bmancini55) [reply lightning/src/routing/network_graph.rs:78]: Hmm.. I'm using the value in the tests so I can drop batch size to 1 and test multipart replies.  Should I add a compile flag to it perhaps?

File: lightning/src/routing/network_graph.rs
Comment (bmancini55) [reply lightning/src/routing/network_graph.rs:377]: I'd like to. Legacy LND replied with the query's `first_blocknum` and `number_of_blocks` using `sync_complete` field to track whether a query was fully replied to, much cleaner that way.  Most implementations recognized this as valid, not sure if they still do. Then LND changed to use implicit sequencing along with with c-lightning and Eclair.  Now we're back to using `sync_complete` to drive sequencing but are still maintaining the sequencing for backwards compatibility reasons. So, long story, sequencing is probably required unless we lobby for a spec change or accept some nodes may not like it.

**Comment by bmancini55** [comment]
Updated with nit fixes!

>  but we will need it to respond to ReplyChannelRange messages, which could be a lot of additional messages and we can't block channel messages while sending them. For that we definitely need to pull the InitSyncTracker struct out into network_graph.rs and make it an associated type for the RoutingMessageHandler trait. I leave it up to you if you want to take that complexity here and also apply it to this code or not.

Do you mean reply to `QueryShortChannelIds`? If so, was thinking the same thing in regards to `InitSyncTracker`, thanks for confirming. I'd like to defer that complexity into the PR that addresses `QueryShortChannelIds`.

File: lightning/src/ln/msgs.rs
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:1693]: Ah, in that case best to call `std::u32::max_value()` otherwise I won't get confused :)

File: lightning/src/routing/network_graph.rs
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:78]: Hmmm, 8000 isn't *that* bad, can we just generate 8k entries in the test?

File: lightning/src/routing/network_graph.rs
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:377]: Hmm, but my point was the spec doesn't actually seem to require it? Nodes might, but my pseudo-careful read of the spec seemed to say you can send the query params in every message.

File: lightning/src/routing/network_graph.rs
Comment (ariard) [inline lightning/src/routing/network_graph.rs:385]: `end_block` isn't mentioned in the spec nor the name of any of our variables, maybe replace by `end_blocknum` ?

**Review by ariard — COMMENTED** [review]
Overall sounds good, mostly minor comments. I did exercise the test coverage, can't find holes. 

> This implementation enqueues each reply message as an event. Right now, at most this will only generate a few events per query due to the graph size. In the future it could generate a large number of events if the graph is large. I'm considering switching to a polling model like the init_sync code in do_attempt_write_data to be more aware of the write buffer. Wanted opinions before I go down that path though.

What are you thinking of more exactly ? I would say the rational for the direction is justified, though can't identify the `init_sync` code you're mentioning.

File: lightning/src/routing/network_graph.rs
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:377]: Right, but equal or greater can be equal :).

**Comment by TheBlueMatt** [comment]
> Do you mean reply to QueryShortChannelIds? If so, was thinking the same thing in regards to InitSyncTracker, thanks for confirming. I'd like to defer that complexity into the PR that addresses QueryShortChannelIds.

Yes, and, ok, sounds good. Can you add a comment noting the amount of memory that can be used in these messages and at what point we should change the logic to avoid it?

File: lightning/src/routing/network_graph.rs
Comment (bmancini55) [reply lightning/src/routing/network_graph.rs:377]: Doh, yes it is! Some day I'll learn logic haha. Will clean this up!

**Comment by bmancini55** [comment]
@ariard thanks for the review! Will clean those up points.  

> I would say the rational for the direction is justified, though can't identify the init_sync code you're mentioning.

I was refer to [`PeerHandler::do_attempt_write_data`](https://github.com/rust-bitcoin/rust-lightning/blob/main/lightning/src/ln/peer_handler.rs#L538) where it uses the `InitSyncTracker` enum to track state of an initial sync.  I was planning to apply this (or similar) to process `QueryShortChannelIds` in another PR.

**Review by TheBlueMatt — APPROVED** [review]

**Review by ariard — COMMENTED** [review]
Code Review ACK 7b842b6

Thanks for simplifying sequencing of `handle_query_channel_range`. Reading the spec again, this new way to fulfill `first_blocknum`/`number_of_blocks` is compliant, all good.

**Pull request merged**
